### ISSUE TITLE
TxnMgr: no PID while aborting

### ISSUE THREAD

**Issue by edenhill** [edenhill/librdkafka#2787]
This surfaced during test 0105 with ENABLE_DEVEL set (thus the dassert):
`Assertion failed: (!*"BUG: No PID despite proper transaction state"), function rd_kafka_txn_op_abort_transaction, file rdkafka_txnmgr.c, line 2227.`

I think we need to handle this better, first understand why there is no PID, and then not raise a retriable error, this is probably a fatal error.

This on v1.4.0-RC5.

**Comment by sanjay24** [comment]
Hi @edenhill,
I got a similar issue on ``v1.5.0``. Unfortunately, I didn't have debug logs enabled during this time. I've the following mix of logs (with my application and rdkafka error logs). 
I'm using transaction in "consume->transform->produce" scenario. This happens at the very start of the application

```
11:08:54.270180313 [Info  ] [MY_PRODUCER] Committing transactions: num-msgs=91088, commit-rate=90870.19
11:08:54.270414294 [Info  ] [MY_PRODUCER] sending 252 top-par offsets to tx coordinator
11:08:54.278628694 [Info  ] [MY_PRODUCER] successfully sent offsets. Committing transaction now!
11:13:54.266699088 [Warn ] [MY_CONSUMER] [anonymous] [4] [MAXPOLL] [thrd:main]: Application maximum poll interval (300000ms) exceeded by 2ms (adjust max.poll.interval.ms for long-running message processing): leaving group
11:13:54.498961021 [Warn ] [MY_CONSUMER] [anonymous] [4] [MAXPOLL] [thrd:main]: Application maximum poll interval (300000ms) exceeded by 228ms (adjust max.poll.interval.ms for long-running message processing): leaving group
11:18:54.141360764 [Warn ] [MY_PRODUCER] [main] failed to commit offsets, error=_STATE: No PID available (idempotence state RequestPID).
 ```

I checked the broker logs and found no error in the logs, 

Here are the problems:

1. This error ``No PID available`` is not logged as abortable/fatal error i.e. rd_kafka_error_txn_requires_abort/rd_kafka_error_is_fatal return false and application continues to move ahead with beginning a new transaction, that obvisously fails with ``error=_STATE: Operation not valid in state BeginCommit``
2. If you look at the logs above there's a 5m interval of no activity, I believe that's because librdkafka is trying to flush messages before committing the transaction. So, it looks like ``rd_kafka_flush`` is stuck forever as it hasn't received back ``InitProducerId`` response. 
3. Any idea why application would fail to get the PID?

Restarting application soon after that works and this issue is kind of very random

**Comment by sanjay24** [comment]
I've added a similar looking issue #3041

**Comment by adscott** [comment]
@edenhill Was this reliably reproducible? Would be keen to understand why this is happening.

**Comment by edenhill** [comment]
Haven't seen this since the txn fixes in 1.6.0
